feat: Recent patient list

robertdown opened the pull request:
This creates a new session array item for the 10 most recent patients that have been opened. The actual user interface is still to be completed, but the underlying functionality now exists.

I know some may argue this should be it's own module, but the overhead of putting it in a new module versus the slim addition to core I feel warrants just putting it in core.

New global to set the max number of Recent Patient, new List to allow admins to customize which fields are set in the session.

bradymiller commented:
hi @robertdown , Definitely nice to have a recent patient selector (this feature has streamlined my workflow in another emr i have used and has saved my on occasion when i needed to lookup a patient that if forgot the details on to search for). Recommend not having it bound to the session (ie. then it gets reset when a user logout of OpenEMR) and instead having a solution that is persistent after logout/login (simple database table that holds the info would suffice).

bradymiller commented:
Would default to 20 on the new global.
Couldn't quite figure out what the list_options were for, but guessing that will be more clear when there is a user interface.

robertdown commented:
The list options control the actual patient data that is returned. Name fields, DOB, etc.

I'll try to move it to the database.

robertdown commented:
Ok @bradymiller bumped the default list to 20, moved storage to DB from $_SESSION, and added the UI to the Patient Finder tab.

Note, there's a typo on this screenshot, it has since been fixed

![image](https://github.com/openemr/openemr/assets/1381170/0fc58fca-98d2-4f4b-bf67-687fbb063c06)

robertdown commented:
Back to draft - I'm gonna Twigify a bunch of stuff

bradymiller commented on interface/main/finder/dynamic_finder.php:
```diff
@@ -373,25 +374,63 @@ function persistCriteria(el, e) {
-                            </tbody>
-                        </table>
+                <ul class="nav nav-tabs mt-3" id="finderTabs" role="tablist">
+                    <li class="nav-item" role="presentation"><button class="nav-link active" id="listTab" data-toggle="tab" data-target="#list" type="button" role="tab" aria-controls="list" aria-selected="true"><?php echo xl("Patient List"); ?></button></li>
```
```echo xlt("Patient List");```

bradymiller commented on interface/main/finder/dynamic_finder.php:
```diff
@@ -373,25 +374,63 @@ function persistCriteria(el, e) {
-                        </table>
+                <ul class="nav nav-tabs mt-3" id="finderTabs" role="tablist">
+                    <li class="nav-item" role="presentation"><button class="nav-link active" id="listTab" data-toggle="tab" data-target="#list" type="button" role="tab" aria-controls="list" aria-selected="true"><?php echo xl("Patient List"); ?></button></li>
+                    <li class="nav-item" role="presentation"><button class="nav-link" id="recentTab" data-toggle="tab" data-target="#recent" type="button" role="tab" aria-controls="recent" aria-selected="true"><?php echo xl("Recent Patients"); ?></button></li>
```
`echo xl("Recent Patients");`

bradymiller commented on interface/main/finder/dynamic_finder.php:
```diff
@@ -373,25 +374,63 @@ function persistCriteria(el, e) {
+                                echo "<thead class=\"thead-dark\"><tr>";
+                                while ($row = sqlFetchArray($res)) {
+                                    $headers[] = $row;
+                                    echo "<th scope=\"col\">" . $row['title'] . "</th>";
```
`text($row['title'])`

bradymiller commented on interface/main/finder/dynamic_finder.php:
```diff
@@ -373,25 +374,63 @@ function persistCriteria(el, e) {
+                                        if ($h['option_id'] == "pid") {
+                                            continue;
+                                        }
+                                        echo "<td><a href=\"#\" onclick=\"javascript:top.RTop.location = '/interface/patient_file/summary/demographics.php?set_pid=" . attr($p['pid']) . "'\">" . $p[$h['option_id']] . "</a></td>";
```
echo "<td><a href=\"#\" onclick=\"javascript:top.RTop.location = '/interface/patient_file/summary/demographics.php?set_pid=" . attr_url($p['pid']) . "'\">" . text($p[$h['option_id']]) . "</a></td>";

bradymiller commented on sql/7_0_1-to-7_0_2_upgrade.sql:
```diff
@@ -228,3 +228,19 @@ INSERT INTO `list_options` (`list_id`, `notes`, `title`, `seq`, `option_id`, `ac
+#IfNotTable recent_patients
+CREATE TABLE recent_patients (
+    user_id varchar(40) NOT NULL,
+    patients TEXT NULL,
```
remove NULL in above line

bradymiller commented on sql/database.sql:
```diff
@@ -13321,3 +13321,15 @@ INSERT INTO `list_options` (`list_id`, `notes`, `title`, `seq`, `option_id`, `ac
+
+CREATE TABLE recent_patients (
+    user_id varchar(40) NOT NULL,
+    patients TEXT NULL,
```
remove NULL in above line

bradymiller commented on src/Services/PatientService.php:
```diff
@@ -963,4 +968,60 @@ private function getPatientSuffixKeys()
+
+        $query = "SELECT patients FROM recent_patients WHERE user_id = ?";
+        $res = sqlStatement($query, $curUser['id']);
+        $row = sqlFetchArray($res);
```
could turn the above 2 lines into following one line:
```
$row = sqlQuery($query, $curUser['id']);
```

bradymiller commented on src/Services/PatientService.php:
```diff
@@ -963,4 +968,60 @@ private function getPatientSuffixKeys()
+        array_unshift($rp, $patient);
+
+        // Cap out at max count as set in globals
+        if (count($rp) == ($GLOBALS['recent_patient_count'] + 1)) {
```
would also do a >= in case it somehow is above (which would require popping the amount over). Could make this happen if go from a larger global to a smaller global and if that is the case then supporting >= would deal with that.

bradymiller commented on src/Services/PatientService.php:
```diff
@@ -963,4 +968,60 @@ private function getPatientSuffixKeys()
+        $sql = "INSERT INTO recent_patients (user_id, patients) VALUES (?, ?) ON DUPLICATE KEY UPDATE patients=?";
+        $res = sqlStatement($sql, [$curUser['id'], $_rp, $_rp]);
+
+        return null;
```
why not just have no return and make the return type  `void` for this function?

robertdown replied:
Because I was writing this function with a migraine.... will set to void

bradymiller commented on src/Services/PatientService.php:
```diff
@@ -963,4 +968,60 @@ private function getPatientSuffixKeys()
+        $currUser = $user->getCurrentlyLoggedInUser();
+        $sql = "SELECT patients FROM recent_patients WHERE user_id = ?";
+        $res = sqlStatement($sql, [$currUser['id']]);
+        $ret = sqlFetchArray($res);
```
could turn the above 2 lines into following one line:
```
$ret = sqlQuery($sql, [$currUser['id']]);
```

bradymiller commented:
Also rec dealing with any php8 warnings when there are no recent patients in the dynamic finder

bradymiller commented:
review complete

bradymiller commented:
note i haven't reviewed the last 2 commits yet :)

robertdown commented:
No recent patients should just return an empty array and display some text indicating there are no patients. Can circle back if needed and delve into the error log

robertdown commented:
Ok @bradymiller addressed the various comments, should be good to go now.

bradymiller commented on interface/patient_file/summary/demographics.php:
```diff
@@ -32,6 +32,7 @@
 use OpenEMR\Billing\EDI270;
 use OpenEMR\Common\Acl\AclMain;
 use OpenEMR\Common\Csrf\CsrfUtils;
+use OpenEMR\Common\Session\PatientSessionUtil;
```
is above line being used in this script?

bradymiller commented on src/Services/PatientService.php:
```diff
@@ -38,6 +38,11 @@ class PatientService extends BaseService
+    /**
+     * @var Array
+     */
+    private $recentPatientArray = [];
```
not seeing above variable used anywhere

bradymiller commented:
hi @robertdown , just noted 2 minor issues. Otherwise code looks great and ready to bring in :)
